This is the first task towards completing the KPMG AU Virtual Internship program. I have been tasked with accessing 3 datasets provided by Sprocket Central Pty Ltd  with for its data quality isuues, the 3 datasets include;
 - Customer Demographic 
 - Customer Addresses
 - Transactions data in the past 3 months

 After this accessment, I am to draft an email on the data quality issues discovered from this exploration stage and ways through which they can be rectified


 ### Exploration of the Customer Demographic Dataset

In [13]:

import pandas as pd

cust_demo_df = pd.read_excel('/KPMG_VI_New_raw_data_update_final.xlsx',
                            sheet_name = 'CustomerDemographic', skiprows = 1)
cust_demo_df.head()

<ipython-input-13-539b4b21080f>:3: FutureWarning: Inferring datetime64[ns] from data containing strings is deprecated and will be removed in a future version. To retain the old behavior explicitly pass Series(data, dtype={value.dtype})
  cust_demo_df = pd.read_excel('/KPMG_VI_New_raw_data_update_final.xlsx',


customer_id      first_name  last_name  gender  \
0            1         Laraine  Medendorp       F   
1            2             Eli    Bockman    Male   
2            3           Arlin     Dearle    Male   
3            4          Talbot        NaN    Male   
4            5  Sheila-kathryn     Calton  Female   

   past_3_years_bike_related_purchases        DOB               job_title  \
0                                   93 1953-10-12     Executive Secretary   
1                                   81 1980-12-16  Administrative Officer   
2                                   61 1954-01-20      Recruiting Manager   
3                                   33 1961-10-03                     NaN   
4                                   56 1977-05-13           Senior Editor   

  job_industry_category     wealth_segment deceased_indicator  \
0                Health      Mass Customer                  N   
1    Financial Services      Mass Customer                  N   
2              Property      Mass Customer                  N   
3                    IT      Mass Customer                  N   
4                   NaN  Affluent Customer                  N   

                                             default owns_car  tenure  
0                                                 "'      Yes    11.0  
1                       <script>alert('hi')</script>      Yes    16.0  
2                                2018-02-01 00:00:00      Yes    15.0  
3  () { _; } >_[$($())] { touch /tmp/blns.shellsh...       No     7.0  
4                                                NIL      Yes     8.0

In [14]:
print(cust_demo_df.shape)
cust_demo_df.info()

(4000, 13)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 13 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   customer_id                          4000 non-null   int64         
 1   first_name                           4000 non-null   object        
 2   last_name                            3875 non-null   object        
 3   gender                               4000 non-null   object        
 4   past_3_years_bike_related_purchases  4000 non-null   int64         
 5   DOB                                  3913 non-null   datetime64[ns]
 6   job_title                            3494 non-null   object        
 7   job_industry_category                3344 non-null   object        
 8   wealth_segment                       4000 non-null   object        
 9   deceased_indicator                   4000 non-null   object        
 10  d

In [15]:
cust_demo_df.isna().sum()

customer_id                              0
first_name                               0
last_name                              125
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     87
job_title                              506
job_industry_category                  656
wealth_segment                           0
deceased_indicator                       0
default                                302
owns_car                                 0
tenure                                  87
dtype: int64

In [16]:
cust_demo_df.gender.value_counts()

Female    2037
Male      1872
U           88
F            1
Femal        1
M            1
Name: gender, dtype: int64

In [17]:
cust_demo_df.deceased_indicator.unique()

array(['N', 'Y'], dtype=object)

In [33]:
cust_demo_df.owns_car.unique()

array(['Yes', 'No'], dtype=object)

In [18]:
cust_demo_df.job_title.unique()

array(['Executive Secretary', 'Administrative Officer',
       'Recruiting Manager', nan, 'Senior Editor', 'Media Manager I',
       'Business Systems Development Analyst', 'Senior Quality Engineer',
       'Nuclear Power Engineer', 'Developer I', 'Account Executive',
       'Junior Executive', 'Media Manager IV', 'Sales Associate',
       'Professor', 'Geological Engineer', 'Project Manager',
       'Safety Technician I', 'Research Assistant I',
       'Accounting Assistant III', 'Editor', 'Research Nurse',
       'Safety Technician III', 'Staff Accountant III', 'Legal Assistant',
       'Product Engineer', 'Information Systems Manager',
       'VP Quality Control', 'Social Worker', 'Senior Cost Accountant',
       'Assistant Media Planner', 'Payment Adjustment Coordinator',
       'Food Chemist', 'Accountant III', 'Director of Sales',
       'Senior Financial Analyst', 'Registered Nurse',
       'Biostatistician II', 'Computer Systems Analyst II',
       'Software Test Engineer II', 

In [20]:
cust_demo_df.job_industry_category.unique()

array(['Health', 'Financial Services', 'Property', 'IT', nan, 'Retail',
       'Argiculture', 'Manufacturing', 'Telecommunications',
       'Entertainment'], dtype=object)

In [22]:
cust_demo_df.wealth_segment.unique()

array(['Mass Customer', 'Affluent Customer', 'High Net Worth'],
      dtype=object)

In [25]:
cust_demo_df.tenure.unique()

array([11., 16., 15.,  7.,  8., 13., 20.,  9.,  6.,  1., 18., 21., 12.,
       19., 14.,  4., 22.,  5., 17.,  2.,  3., 10., nan])

In [28]:
cust_demo_df.past_3_years_bike_related_purchases.unique()

array([93, 81, 61, 33, 56, 35,  6, 31, 97, 49, 99, 58, 38, 85, 91, 57, 79,
       76, 72, 74, 43, 55, 12, 37,  5, 62, 18,  3, 17, 44, 59, 40, 46, 64,
       24, 63, 98, 51, 68, 22,  2, 48, 87, 26, 60, 47, 73, 21, 67, 16, 78,
       30, 94, 28, 20, 11,  1, 75, 41, 69, 19, 80, 83, 25, 54,  4, 23, 65,
       88, 10, 77,  9, 82, 27, 53, 32, 34, 71, 36, 90, 95,  8, 39,  7, 42,
       13, 45, 50, 14, 89, 84, 96, 70, 66,  0, 15, 86, 92, 29, 52])

In [29]:
cust_demo_df.duplicated().sum()

0

In [31]:
cust_demo_df.default.unique()

array(['"\'', "<script>alert('hi')</script>",
       datetime.datetime(2018, 2, 1, 0, 0),
       '() { _; } >_[$($())] { touch /tmp/blns.shellshock2.fail; }',
       'NIL', 'ðµ ð ð ð', 'â°â´âµâââ', '(â¯Â°â¡Â°ï¼â¯ï¸µ â»ââ»)', '0/0',
       'ð©ð½', 'Åâ´®â\xa0¥¨ËøÏââ', 'nil', -100, 'â°â´âµ', 'ð', 1e+96,
       'ï¾ï½¥â¿ã¾â²(｡ââ¿â｡)â±â¿ï½¥ï¾', 'Ωâçââ«Ëµâ¤â¥Ã·',
       'Åâ´â°ËèËÃâââ', 'ï¼ï¼ï¼',
       '../../../../../../../../../../../etc/hosts',
       '×Ö¸×ְתָ×testاÙصÙØ\xadات اÙتÙØ\xadÙÙ', '<>?:"{}|_+',
       '\'\'\'\'"', ",./;'[]\\-=",
       '() { 0; }; touch /tmp/blns.shellshock1.fail;',
       'ì¬íê³¼íì ì´íì°êµ¬ì', 'testâ\xa0testâ«',
       '0ï¸â£ 1ï¸â£ 2ï¸â£ 3ï¸â£ 4ï¸â£ 5ï¸â£ 6ï¸â£ 7ï¸â£ 8ï¸â£ 9ï¸â£ ð',
       nan, '!@#$%^&*()', "'",
       '̦HÍ̬̤Ì̤eÍ ÍÌÌ¥ÌÌ»ÍÌwÌhÌ̯ÍoÌÍÌÍ̱̮ Ò̺ÌÌÌÍW̷̼Ì\xada̺̪ÍįÍÍÌ\xadÍ̯Ìt̶̼̮sÌÌÍÍ Ì\xa0Ì«Ì\xa0BÌ»ÍÍÍÍ̳e̵h̵̬ÍÌ«ÍiÌ̹Í̳̳̮ÍÌ«nÍd̴̪ÌÌ ÍÌ°ÍÌ©ÍÍÍ̲TÍ¢Í̼Í̪hÍÍ̮̻e̬ÌÍÌ Ì¤Ì¹ÌWÍÍÍÌÌÍÍaÍÍÍ̹̼',
       '../../../

In [32]:
cust_demo_df.columns

Index(['customer_id', 'first_name', 'last_name', 'gender',
       'past_3_years_bike_related_purchases', 'DOB', 'job_title',
       'job_industry_category', 'wealth_segment', 'deceased_indicator',
       'default', 'owns_car', 'tenure'],
      dtype='object')

In [35]:
cust_demo_df['birth_year'] = cust_demo_df['DOB'].dt.year
cust_demo_df.birth_year.nunique()

56

In [36]:
cust_demo_df['age'] = 2023 - cust_demo_df['birth_year']
cust_demo_df.age.unique()

array([ 70.,  43.,  69.,  62.,  46.,  57.,  47.,  61.,  50.,  35.,  29.,
        68.,  40.,  23.,  56.,  22.,  28.,  45.,  55.,  66., 180.,  60.,
        38.,  44.,  42.,  48.,  51.,  31.,  64.,  67.,  27.,  53.,  58.,
        21.,  63.,  49.,  54.,  30.,  34.,  37.,  65.,  39.,  25.,  26.,
        33.,  24.,  41.,  59.,  36.,  nan,  52.,  32.,  79.,  92.,  88.,
        80.,  83.])

In [37]:
cust_demo_df.query('age == 180')

customer_id first_name last_name gender  \
33           34   Jephthah  Bachmann      U   

    past_3_years_bike_related_purchases        DOB        job_title  \
33                                   59 1843-12-21  Legal Assistant   

   job_industry_category     wealth_segment deceased_indicator default  \
33                    IT  Affluent Customer                  N     NaN   

   owns_car  tenure  birth_year    age  
33       No    20.0      1843.0  180.0

In the customer Demographics, after the data quality assessment, the following data quality issues were discoverd;

 - Completeness
 
 The dataset is made up of 4000 entries with 13 attributes (columns).
 The dataset contains missing values in the following columns;
 __last_name - 125 missing values__,
 __DOB - 87 missing values__,
 __job_title - 506 missing values__,
 __job_industry_category - 656 missing values__,
 __default - 302 missing values__,
 __tenure - 87 missing values__.

 **Completeness Solution*

 1. We can fill in the missing values in the categorical variables like the job_industry_category with the most occurring category

 2. For the the missing values in the DOB column, we can compute the age from using the DOB column and then fill in the missing values with the mean age.

 3. the default column does not contain any information relevant to the analysis of the data and so will be dropped.

 4. Additional information on the column, tenure is needed inorder for me to know the appropriate way to tackle the missingness in that column.

 5. The rest of the missing values will be dropped.





- Consistency

The __gender__ column is made up of 3 distinct genders, Male, Female, and Unknown. But the representations of these genders are not consistent. Also, the representation of the 'Yes' and 'No' throughout the dataset is not consistent. In the __deceased_indicator__ column, it is represented using the first letter uppercase ('Y', 'N'), while in the __owns_car__ column, it is represented in full ('Yes', 'No').

*Consistency Solution*

1. The incorrectly spelt gender eg *'Female'* will be replaced with the correct spelling.

2. The gender designation not represented in full eg (*'U'*,*'F'*) will be spelt in full.

3. The 'Y' and 'N' designation used to represent 'Yes' and 'No' in the deceased_indicator column will be spelt in full



 - Relevancy

 The __default__ column contains data that are not relevant, it is not human readable and therefore doesn't add any useful information that could be relevant to the analysis of this data set.

*Relevancy Solution*

The default column will be dropped


 - Validity

 A customer, Jephthah 	Bachmann 	according to the dataset, is born on 1843-12-21 which means that this customer is 180years, this could probably be some mistake.

The __customer_id__ is of an invalid datatype(int). The customer_id serves as a means of identification for the customers and so should be of the 'object' data type.

*Validity Solution*

1. The customer's date of birth will be replaced with 1943-12-21

2. The customer_id dtype will be converted to 'object'



## Transaction data set


In [48]:
# load the data set
trans_df = pd.read_excel('/KPMG_VI_New_raw_data_update_final.xlsx',
                            sheet_name = 'Transactions', skiprows = 1)
trans_df.head()

transaction_id  product_id  customer_id transaction_date  online_order  \
0               1           2         2950       2017-02-25           0.0   
1               2           3         3120       2017-05-21           1.0   
2               3          37          402       2017-10-16           0.0   
3               4          88         3135       2017-08-31           0.0   
4               5          78          787       2017-10-01           1.0   

  order_status           brand product_line product_class product_size  \
0     Approved           Solex     Standard        medium       medium   
1     Approved   Trek Bicycles     Standard        medium        large   
2     Approved      OHM Cycles     Standard           low       medium   
3     Approved  Norco Bicycles     Standard        medium       medium   
4     Approved  Giant Bicycles     Standard        medium        large   

   list_price  standard_cost  product_first_sold_date  
0       71.49          53.62                  41245.0  
1     2091.47         388.92                  41701.0  
2     1793.43         248.82                  36361.0  
3     1198.46         381.10                  36145.0  
4     1765.30         709.48                  42226.0

In [39]:
# get an overview of the data
print(trans_df.shape)
trans_df.info()

(20000, 13)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   transaction_id           20000 non-null  int64         
 1   product_id               20000 non-null  int64         
 2   customer_id              20000 non-null  int64         
 3   transaction_date         20000 non-null  datetime64[ns]
 4   online_order             19640 non-null  float64       
 5   order_status             20000 non-null  object        
 6   brand                    19803 non-null  object        
 7   product_line             19803 non-null  object        
 8   product_class            19803 non-null  object        
 9   product_size             19803 non-null  object        
 10  list_price               20000 non-null  float64       
 11  standard_cost            19803 non-null  float64       
 12  product_first_sold_d

In [40]:
# check for missing values
trans_df.isna().sum()

transaction_id               0
product_id                   0
customer_id                  0
transaction_date             0
online_order               360
order_status                 0
brand                      197
product_line               197
product_class              197
product_size               197
list_price                   0
standard_cost              197
product_first_sold_date    197
dtype: int64

In [44]:
# checking for unique values inorder to make sure there is no incorrect entry
print(trans_df.brand.unique())

['Solex' 'Trek Bicycles' 'OHM Cycles' 'Norco Bicycles' 'Giant Bicycles'
 'WeareA2B' nan]


In [43]:
print(trans_df.product_line.unique())
print(trans_df.product_class.unique())
print(trans_df.product_size.unique())
print(trans_df.order_status.unique())

['Standard' 'Road' 'Mountain' 'Touring' nan]
['medium' 'low' 'high' nan]
['medium' 'large' 'small' nan]
['Approved' 'Cancelled']


In [50]:
# checking the transaction_date to make sure there are no irregularities in the transaction date column
trans_df.transaction_date.unique()

array(['2017-02-25T00:00:00.000000000', '2017-05-21T00:00:00.000000000',
       '2017-10-16T00:00:00.000000000', '2017-08-31T00:00:00.000000000',
       '2017-10-01T00:00:00.000000000', '2017-03-08T00:00:00.000000000',
       '2017-04-21T00:00:00.000000000', '2017-07-15T00:00:00.000000000',
       '2017-08-10T00:00:00.000000000', '2017-08-30T00:00:00.000000000',
       '2017-01-17T00:00:00.000000000', '2017-01-05T00:00:00.000000000',
       '2017-02-26T00:00:00.000000000', '2017-09-10T00:00:00.000000000',
       '2017-06-11T00:00:00.000000000', '2017-10-10T00:00:00.000000000',
       '2017-04-03T00:00:00.000000000', '2017-06-02T00:00:00.000000000',
       '2017-04-06T00:00:00.000000000', '2017-01-28T00:00:00.000000000',
       '2017-10-09T00:00:00.000000000', '2017-06-29T00:00:00.000000000',
       '2017-04-08T00:00:00.000000000', '2017-10-18T00:00:00.000000000',
       '2017-01-10T00:00:00.000000000', '2017-04-11T00:00:00.000000000',
       '2017-12-23T00:00:00.000000000', '2017-10-13

In [52]:
trans_df['product_first_sold_date'] = pd.to_datetime(trans_df['product_first_sold_date'], format='%Y%m%d')
trans_df['product_first_sold_date'].head()

0   1970-01-01 00:00:00.000041245
1   1970-01-01 00:00:00.000041701
2   1970-01-01 00:00:00.000036361
3   1970-01-01 00:00:00.000036145
4   1970-01-01 00:00:00.000042226
Name: product_first_sold_date, dtype: datetime64[ns]

In [53]:
trans_df.online_order.unique()

array([ 0.,  1., nan])

In [58]:
trans_df.duplicated().sum()

0

In the Transactions Dataset, after the data quality assessment, the following data quality issues were discoverd;

 - Completeness
 
 The dataset is made up of 2000 entries with 13 attributes (columns).
 The dataset contains missing values in the following columns;
 __Onlin_order - 360 missing values__,
 __brand - 197 missing values__,
 __product_line - 197 missing values__,
 __product_class - 197 missing values__,
 __product_size - 197 missing values__,
 __standard_cost - 197 missing values__.

*Solution*

1. First, I will get in contact with the company to know if there is anyway this could be rectified by providing an additional information which will enable me fill in the missing values. If not, Most of the data completeness issues are found in categorical variables and thus could be filled using the most occurring. The missing values in the standard cost could be fill in using the mean standard cost, and the rest of the missing values will be dropped.

- Consistency

The Data Representations accross the data set are not consistent, in some columns. data entries are in the title case while some others are represented in lower case.

*Solution*

All cases will be converted to the lower case

 - Accuracy

The following columns __transaction_id__, __product_id__, and __customer_id__ are of incorrect dtype (int). These columns are used in identification of an observation and no mathematical operation is expected to be performed on these columns, and so should be of the type 'str' or 'object'. Also, the following column, __product_line__, __product_class__, __product_size__,__standard_cost__ should be of the dtype, 'category'.

*Solution*

I will change the datatypes using the .astype() method

 - Relevancy

The column, __product_first_sold_date__ is of the wrong dtype(float), and does not contain the relevant information needed in this analysis

*Solution*

I will request on additional information that will help me fix the column in order to provide useful information for the analysis. When none is provided, the column will be dropped. 

## Customer Addresses Data Set


In [54]:
# load the data set
address_df = pd.read_excel('/KPMG_VI_New_raw_data_update_final.xlsx',
                            sheet_name = 'CustomerAddress', skiprows = 1)
address_df.head()

customer_id              address  postcode            state    country  \
0            1   060 Morning Avenue      2016  New South Wales  Australia   
1            2  6 Meadow Vale Court      2153  New South Wales  Australia   
2            4   0 Holy Cross Court      4211              QLD  Australia   
3            5  17979 Del Mar Point      2448  New South Wales  Australia   
4            6     9 Oakridge Court      3216              VIC  Australia   

   property_valuation  
0                  10  
1                  10  
2                   9  
3                   4  
4                   9

In [57]:
print(address_df.shape)
address_df.info()

(3999, 6)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3999 entries, 0 to 3998
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   customer_id         3999 non-null   int64 
 1   address             3999 non-null   object
 2   postcode            3999 non-null   int64 
 3   state               3999 non-null   object
 4   country             3999 non-null   object
 5   property_valuation  3999 non-null   int64 
dtypes: int64(3), object(3)
memory usage: 187.6+ KB


In [59]:
address_df.country.unique()

array(['Australia'], dtype=object)

In [60]:
address_df.duplicated().sum()

0

## Address Dataset Quality Issues


 - __customer_id__ should be of dtype('object') not int
 - __postcode__ should be of dtype('object') not int
 - The data representation in the __state__ column is not consistent
 - Since all observations originated from Australia, the __country__ column is not relevant.

*Solution* 
 
 Rectify them as I did above.